util/EIP-712/src/encode.rs

```diff
@@ -0,0 +1,339 @@
+		temp.insert(0, message_type);
+		temp
+	};
+
+	let encoded = deps
+		.into_iter()
+		.filter_map(|dep| {
+			message_types.get(dep).map(|field_types| {
+				let types = field_types
+					.iter()
+					.map(|value| format!("{} {}", value.type_, value.name))
+					.join(",");
+				return format!("{}({})", dep, types);
+			})
+		})
+		.collect::<Vec<_>>()
+		.concat();
+	Ok(encoded)
+}
+
+fn type_hash(message_type: &str, typed_data: &MessageTypes) -> Result<H256> {
+	Ok(keccak(encode_type(message_type, typed_data)?))
+}
+
+fn encode_data(
+	parser: &Parser,
+	message_type: &Type,
+	message_types: &MessageTypes,
+	value: &Value,
+	field_name: Option<&str>
+) -> Result<Vec<u8>>
+{
+	let encoded = match message_type {
+		Type::Array(array_type) => {
+			let mut items = vec![];
+
+			for item in value.as_array().ok_or_else(|| serde_error("array", field_name))? {
+				let mut encoded = encode_data(parser, &*array_type, &message_types, item, field_name)?;
+				items.append(&mut encoded);
+			}
+
+			keccak(items).0.to_vec()
+		}
+
+		Type::Custom(ref ident) if message_types.get(&*ident).is_some() => {
+			let type_hash = (&type_hash(ident, &message_types)?).to_vec();
+			let mut tokens = encode(&[EthAbiToken::FixedBytes(type_hash)]);
+
+			for field in message_types.get(ident).expect("Already checked in match guard; qed") {
+				let value = &value[&field.name];
+				let type_ = parser.parse_type(&*field.type_)?;
+				let mut encoded = encode_data(parser, &type_, &message_types, &value, Some(&*field.name))?;
+				tokens.append(&mut encoded);
+			}
+
+			keccak(tokens).0.to_vec()
+		}
+
+		Type::Bytes(size) => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() < 2 {
+				return Err(ErrorKind::HexParseError(
+					format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+				)?
+			}
+			let mut bytes = H256::from_str(&string[2..]).map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			if *size == 32 {
+				bytes = keccak(&bytes);
+			}
+			encode(&[EthAbiToken::FixedBytes(bytes.to_vec())])
+		}
+
+		Type::String => {
+			let value = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			let hash = (&keccak(value)).to_vec();
+			encode(&[EthAbiToken::FixedBytes(hash)])
+		}
+
+		Type::Bool => encode(&[EthAbiToken::Bool(value.as_bool().ok_or_else(|| serde_error("bool", field_name))?)]),
+
+		Type::Address => {
+			let addr = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if addr.len() != 42 {
+				return Err(ErrorKind::InvalidAddressLength(addr.len()))?;
+			}
+			let address = EthAddress::from_str(&addr[2..]).map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			encode(&[EthAbiToken::Address(address)])
+		}
+
+		Type::Uint | Type::Int => {
+			// try to deserialize as a number first, then a string
+			let uint = match (value.as_u64(), value.as_str()) {
+				(Some(number), _) => U256::from(number),
```

Comment:
Rised that earlier already - are we sure we want to support amibgious formats for numbers (i.e. regular numbers and hex-encoded strings?) The reason we don't do that in RPC is because JavaScript doesn't support large numbers natively, and accepting numbers fools the developers that they can just pass arbitrary number (even U256) to the API and it will work, while JS might decide differently, so it's very error-prone.
I'm strongly in favor of supporting only hex-encoded (`QUANTITY`) numbers

Reply:
Yeah, you're right, will remove it.